### DISCUSSION TITLE
How to use `Result.transpiledFiles`?

### DISCUSSION THREAD

**Discussion by KOLANICH** [google/closure-compiler#4064] [Q&A] [answered]
[Here is my code](https://github.com/KOLANICH-libs/ClosureCompiler.py/blob/master/ClosureCompiler/__init__.py), but it seems that `Result.transpiledFiles` are always empty (in my scenario, see [the test](https://github.com/KOLANICH-libs/ClosureCompiler.py/blob/master/tests/tests.py)).

Should I get the results into `Result.transpiledFiles` instead of `compiler.toSource()` and what should I do for it?

**Comment by niloc132** [comment]
It might be easier if you started with a Java program, rather than bridging between python and Java.

In my experience, the command line options are deceptively simple compared to how flexible the Compiler class is. I've found it is nearly impossible to use it without first starting from the command line runner, figuring out what kinds of options it would have created, and what each of those options would have done to the Compiler to produce what you expected, and only they start tweaking things in ways that are impossible with the command line...

With that said, poking quickly at the Java sources where result.transppiled files is read, for example:
https://github.com/google/closure-compiler/blob/027f5e646ce028ccc45267625403e439dcb8c1bc/src/com/google/javascript/jscomp/transpile/BaseTranspiler.java#L131-L150
and (this one indirectly):
https://github.com/google/closure-compiler/blob/027f5e646ce028ccc45267625403e439dcb8c1bc/src/com/google/javascript/jscomp/bundle/CompilerBasedTransformer.java#L96-L112

In both cases, the compiler isn't optimizing, but just transpiling - as the name suggests. As such, you probably don't want this, if your implementation of createCompiler options is what you actually want
https://github.com/KOLANICH-libs/ClosureCompiler.py/blob/a6ffe19fd7305469ec612267a2308e88762849f4/ClosureCompiler/lowLevel.py#L187

As I understand it, this is mostly/only used for BUNDLE code, to rewrite es6 modules into something that can be inlined into a single output JS file.

In contrast, you're probably after either a single output file, or the initial file and its corresponding chunks? If so, looking at how the command line runner walks the specified chunks then calls `compiler.getSourceFileByName(name)` for each. Note that getSourceFileByName is package-protected, so you probably would use `getSourceFileContentByName` instead for that purpose.

`Compiler.toSource()` is what the command line runner calls on at least a few code paths, so could well be what you actually want to call.

**Comment by KOLANICH** [comment, marked as answer]
>I've found it is nearly impossible to use it without first starting from the command line runner, figuring out what kinds of options it would have created, and what each of those options would have done to the Compiler to produce what you expected, and only they start tweaking things in ways that are impossible with the command line...

Yeah, I did the same.

>poking quickly at the Java sources where `result.transpiledFiles` is read

I have also poked, and found nothing useful. I mean I have tried different options set in those context, but still get an empty array.

But thanks for the info.

It feels like that prop is used only when outputting legacy versions of ECMAScript is requested. The following snippet 

```python
from ClosureCompiler import *
from ClosureCompiler.lowLevel import *

c = Compiler(createCompilerOptions(
    languageOut = ji.CompileLanguageMode.ECMASCRIPT5,
))
res = c({
    "a.jsm": "export const asdf = 'aaaa'; alert('1 ' + asdf);",
})
r = c.compiler.getResult()
s = list(r.transpiledFiles)[0]
print("s.getCode()", repr(str(s.getCode())))
print("res.files[None]", repr(res.files[None]))  # res is my object
```

causes `transpiledFiles` to be populated, but with vanilla source code:

```
s.getCode() "export const asdf = 'aaaa'; alert('1 ' + asdf);"
res.files[None] '\'use strict\';\nalert("1 aaaa");\n'
```

so `transpiledFiles` are not results, but inputs. In fact they are the same objects:

<details>
<summary>reproducing code</summary>

```python
# https://repo.maven.apache.org/maven2/org/openjdk/jol/jol-core/0.17/jol-core-0.17.jar was added into the classpath and "org.openjdk.jol.vm.VM" was added into imported classes

from ClosureCompiler.lowLevel import *

options = createCompilerOptions(
    languageOut = ji.CompileLanguageMode.ECMASCRIPT5,
)

sources = convertSourcesMapping({
    "a.jsm": "export const asdf = 'aaaa'; alert('1 ' + asdf);",
})

builtinExterns = getBuiltInExterns(options.getEnvironment())

compiler = ji.Compiler()
lowLevelRes = compiler.compile(
    ji.ArrayList(builtinExterns),
    ji.ArrayList(sources),
    options,
)

if not lowLevelRes.success:
    raise Exception("Transpilation has failed, see the cli output for now")

s = list(lowLevelRes.transpiledFiles)[0]
vm = ji.VM.current()
vm.addressOf(s) == vm.addressOf(sources[0]) # True
```

</details>

>In contrast, you're probably after either a single output file, or the initial file and its corresponding chunks?

In fact I assumed that the Closure Compiler optimizes each input taking in account the info from other inputs (kinda like externs do, but preserving only keeping the stuff that is determined to be used), but keeping the results a bit separately in different `com.google.javascript.jscomp.SourceFile`s. It seems that currently Closure Compiler doesn't do it itself, but it can probably be solved by multiple compiling calls with compiling one file at time and keeping the rest as externs.

**Comment by niloc132** [comment]
> It feels like that prop is used only when outputting legacy versions of ECMAScript is requested. 

Yes, that is what I was trying to convey. Odds are, you don't want this, _unless_ you have es6 module inputs, and, without optimizing them, want to output them directly. For example, compilationLevel=BUNDLE will combine all JS files into one without any optimizations, and in order to achieve that any `import`s and `export`s need to be rewritten into references to other files that have been sent.

> In fact I assumed that the Closure Compiler optimizes each input taking in account the info from other inputs (kinda like externs do, but preserving only keeping the stuff that is determined to be used), but keeping the results a bit separately in different com.google.javascript.jscomp.SourceFiles.

I don't think you would want this. This would prevent inlining from one file to another from occurring, among other things. The compiler doesn't stop at "treeshaking", but is capable of substantially more than that, which blurs the lines between files.

**Comment by KOLANICH** [comment]
>I don't think you would want this. This would prevent inlining from one file to another from occurring, among other things. The compiler doesn't stop at "treeshaking", but is capable of substantially more than that, which blurs the lines between files.

Yeah, I know. I just assumed it having such a capability, so wanted to expose it through my bindings too.
